class ListNode{
        int val;
        ListNode next;
        ListNode prev;
        public ListNode(int val, ListNode next, ListNode prev){
            this.val = val;
            this.prev = prev;
            this.next = next;
        }
    }

class MyCircularQueue {

    public int space;
    public ListNode left, right;

    public MyCircularQueue(int k) {
        space = k;
        left = new ListNode(0, null, null);
        right = new ListNode(0, null, left);
        left.next = right;
        
    }
    
    public boolean enQueue(int value) {
        if(isFull()){
            return false;
        }
        ListNode cur = new ListNode(value, right, right.prev);
        right.prev.next = cur;
        right.prev = cur;
        space--;
        return true;
    }
    
    public boolean deQueue() {
        if(isEmpty()){
            return false;
        }
        left.next = left.next.next;
        left.next.prev = left;
        space++;
        
        return true;
    }
    
    public int Front() {
        return isEmpty() ? -1 : left.next.val;
    }
    
    public int Rear() {
        return isEmpty() ? -1 : right.prev.val;
    }
    
    public boolean isEmpty() {
         return left.next == right;
    }
    
    public boolean isFull() {
        return space == 0;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */